File: code/swift/StringManipulation.swift
// Strings in Swift are collections of Characters, not of bytes, which is
// why indexing works the way it does.

import Foundation

let line = "Alder Cross,Amber,2,true"
let name = "Quill Wharf"

print("count:", name.count)
print("upper:", name.uppercased(), "| lower:", name.lowercased())
print("capitalized:", "quill moor station".capitalized)

print("contains:", line.contains("Amber"))
print("hasPrefix/hasSuffix:", line.hasPrefix("Alder"), line.hasSuffix("true"))
print("isEmpty:", "".isEmpty)

// Splitting and joining.
let fields = line.split(separator: ",").map(String.init)
print("\(fields.count) fields, last \(fields.last ?? "-")")
print("joined:", fields.prefix(3).joined(separator: " | "))
print("keeping empties:", "a,,b".split(separator: ",", omittingEmptySubsequences: false).count)

// String indices are opaque, because a Character can span several bytes.
let start = name.startIndex
let fifth = name.index(start, offsetBy: 5)
print("prefix:", name[start..<fifth])
print("suffix:", name[fifth...])
print("first/last:", name.first as Any, name.last as Any)
print("prefix(5):", name.prefix(5), "| suffix(5):", name.suffix(5))

if let comma = line.firstIndex(of: ",") {
    print("up to the comma:", line[..<comma])
    print("after it:", line[line.index(after: comma)...])
}

// Replacing and trimming.
print("replaced:", line.replacingOccurrences(of: ",", with: "; "))
print("trimmed: [\("   spaced out   ".trimmingCharacters(in: .whitespaces))]")
print("dropped:", line.dropFirst(12))

// Multibyte text: characters, unicode scalars, and UTF-8 bytes are three
// different counts.
let unicode = "café naïve"
print("characters \(unicode.count), scalars \(unicode.unicodeScalars.count), utf8 \(unicode.utf8.count)")
print("first four:", String(unicode.prefix(4)))

// A combining sequence is one Character even though it is several scalars.
let combined = "e\u{0301}"
print("\(combined) is \(combined.count) character of \(combined.unicodeScalars.count) scalars")

// Building strings.
var report = ""
for zone in 1...5 {
    report += "zone \(zone)"
    if zone < 5 { report += " -> " }
}
print(report)

print(String(repeating: "-", count: 24))
print(String(format: "%@ %5.2f %03d", "Alder Cross" as NSString, 3.4, 2))
print("padded: |\(name.padding(toLength: 20, withPad: " ", startingAt: 0))|")

// Multi-line literals keep the layout and strip the common indentation.
let block = """
    Station: \(name)
    Zone:    2
    """
print(block)

// Raw strings need no escaping, which suits regular expressions and paths.
print(#"a raw string keeps \n and "quotes" as they are"#)

// Parsing and formatting numbers.
print("parsed:", Int("2") as Any, "| bad parse:", Int("east") as Any)
print("formatted:", 1234567.891.formatted(.number.precision(.fractionLength(2))))

// A slug, using only the standard library.
let slug = line.lowercased()
    .map { $0.isLetter || $0.isNumber ? $0 : "-" }
    .reduce(into: "") { result, character in
        if character == "-" && result.hasSuffix("-") { return }
        result.append(character)
    }
print("slug:", slug.trimmingCharacters(in: CharacterSet(charactersIn: "-")))

File: code/ruby/string_manipulation.rb
# Strings: interpolation, the methods worth knowing, and mutation.

line = 'Alder Cross,Amber,2,true'
name = 'Quill Wharf'

puts "interpolated: #{name} has #{name.split.size} words"
puts "length #{line.length}, bytes #{line.bytesize}"
puts line.upcase
puts 'quill moor station'.split.map(&:capitalize).join(' ')

puts "includes Amber: #{line.include?('Amber')}"
puts "starts with: #{line.start_with?('Alder')}, ends with: #{line.end_with?('true')}"
puts "index of comma: #{line.index(',')}"

fields = line.split(',')
puts "#{fields.size} fields, last is #{fields.last}"
puts fields.first(3).join(' | ')

puts line.tr(',', ';')
puts line.sub('Amber', 'Cobalt')
puts line.gsub(/,/, ' / ')
puts line.delete('aeiou')
puts line.squeeze('s')

puts "padded: |#{'left'.ljust(12)}|#{'right'.rjust(12)}|#{'mid'.center(11, '.')}|"
puts "trimmed: [#{'   spaced out   '.strip}]"
puts "sliced: #{line[0, 11]} / #{line[-4..]} / #{line[/[A-Z]\w+/]}"

puts format('%-14s %5.2f %03d', 'Alder Cross', 3.4, 2)
puts '%s is in zone %d' % [name, 3]

# Heredocs: <<~ strips the leading indentation, <<~'X' skips interpolation.
zone = 2
puts(<<~REPORT)
  Station: #{name}
  Zone:    #{zone}
REPORT

# Strings are mutable unless frozen; << changes in place, + makes a copy.
buffer = +'zone '
5.times { |n| buffer << (n + 1).to_s << ' ' }
puts buffer.strip

frozen = 'literal'.freeze
begin
  frozen << '!'
rescue FrozenError => e
  puts "frozen: #{e.message}"
end

# Multibyte strings count characters, not bytes.
unicode = 'café naïve'
puts "characters #{unicode.length}, bytes #{unicode.bytesize}"
puts unicode.chars.first(4).join
puts unicode.unicode_normalize(:nfd).length

puts 'each_char: ' + 'amber'.each_char.to_a.inspect
puts 'scan: ' + line.scan(/[A-Z]\w+/).inspect
puts 'succ: ' + 'ST-001'.succ
puts 'reverse: ' + name.reverse
puts 'slug: ' + line.downcase.gsub(/[^a-z0-9]+/, '-').delete_prefix('-').delete_suffix('-')
